### PR TITLE
Allow switches to start from NpT trajectories

### PR DESCRIPTION
**Pull request by wiederm**
## Description

Allow switches starting from pool of conformations that have varying box vectors.
In the process of preparing for these changes we also realized that fields in the result/protocol dataclass don't use clear language to specify the direction of the switching and what pool of samples is used. This is also addressed in this PR.

## Todos
  - [x] change function signature to allow passing of a mdtraj.Trajectory instance from which switches are initialized. This has the benefit that we have access to the unit cell information if needed.
  - [x] update fields in dataclasses to indicate direction of switching and what pool of samples is used 

## Status
- [ ] Ready to go

### CODE REVIEW

**Comment by wiederm** [comment]
@xiki-tempula , can you take a look at the test that is failing? It's in the part of the code that interactions with BSS.
As soon as it is passing I am happy to merge this PR.

File: endstate_correction/neq.py [hunk 1]
```diff
@@ -11,18 +11,18 @@
...
-    nr_of_switches: int = 50, 
+    sim,
+    lambdas: list,
+    samples: list,
+    nr_of_switches: int = 50,
     save_trajs: bool = False,
-    save_endstates:bool = False,
+    save_endstates: bool = False,
 ) -> Tuple[list, list, list]:
     """performs NEQ switching using the lambda sheme passed from randomly dranw samples"""
```
Comment (xiki-tempula) [inline endstate_correction/neq.py:24]: Do you mind add a doc string here to show what are these variables, please? particularly what is samples. Thank you!

File: endstate_correction/neq.py [hunk 2]
```diff
@@ -49,23 +49,21 @@ def perform_switching(
...
 
         # select a random sample
-        x = (
-            np.array(random.choice(samples).value_in_unit(distance_unit))
-            * distance_unit
-        )
+        random_frame = random.randint(0, len(samples.xyz)-1)
+        x = samples.xyz[random_frame]
+        if samples.unitcell_lengths is not None:
+            box_length = samples.unitcell_lengths[x]
```
Comment (xiki-tempula) [inline endstate_correction/neq.py:58]: It is a bit confusing here as to what this `x` is.
From `x = samples.xyz[random_frame]`, I would imagine that this x would be the coordinate of frame `random_frame` extracted from `samples.xyz`.
Then I would imagine that the box dimension of frame `random_frame` would be `samples.unitcell_lengths[random_frame]`, but this is not the case, so it might be better to explain what is this `x`?

File: endstate_correction/protocol.py [hunk 3]
```diff
@@ -124,11 +127,10 @@ class Results:
     dE_qml_to_mm: np.array = np.array([])
     W_mm_to_qml: np.array = np.array([])
     W_qml_to_mm: np.array = np.array([])
-    endstate_samples_mm_to_qml : np.array = np.array([])
-    endstate_samples_qml_to_mm : np.array = np.array([])
-    switching_traj_mm_to_qml : np.array = np.array([])
-    switching_traj_qml_to_mm : np.array = np.array([])
-   
+    endstate_samples_mm_to_qml: np.array = np.array([])
```
Comment (xiki-tempula) [inline endstate_correction/protocol.py:130]: I think it might be good to move the reformatting to another PR as the diff is not very helpful in this case.

File: endstate_correction/protocol.py [hunk 4]
```diff
@@ -261,26 +268,26 @@ def perform_endstate_correction(protocol: Protocol) -> Results:
...
-                    lambs,
-                    samples=protocol.trajectories[-1],
-                    nr_of_switches=protocol.nr_of_switches,
-                    save_endstates=protocol.save_endstates,
-                    save_trajs=protocol.save_trajs
-                )
-            Ws_from_qml_to_mm = np.array(Ws/ kBT)   
+                sim,
+                lambs,
+                samples=protocol.trajectories[1],
```
Comment (xiki-tempula) [inline endstate_correction/protocol.py:285]: Why the change from `samples=protocol.trajectories[-1]` to `samples=protocol.trajectories[1]`

File: endstate_correction/protocol.py [hunk 5]
```diff
@@ -296,39 +303,39 @@ def perform_endstate_correction(protocol: Protocol) -> Results:
...
```
Comment (xiki-tempula) [inline endstate_correction/protocol.py:330]: Same here. -1 makes sense as one is taking the last frame but I don't understand why you need to take the second frame

File: endstate_correction/tests/test_sampling.py [hunk 6]
```diff
@@ -36,14 +34,11 @@ def test_sampling():
         f"{hipen_testsystem}/{system_name}/{system_name}.str",
     )
     # define region that should be treated with the qml
-    chains = list(psf.topology.chains())
-    ml_atoms = [atom.index for atom in chains[0].atoms()]
-
-    sim = create_charmm_system(
-        psf=psf, parameters=params, env="vacuum", ml_atoms=ml_atoms
-    )
+    sim = setup_vacuum_simulation(psf, params)
```
Comment (xiki-tempula) [inline endstate_correction/tests/test_sampling.py:37]: I think it might be good to separate the code that is not relevant to the switching to NpT traj to another PR>

File: endstate_correction/protocol.py (hunk 4 above)
Comment (wiederm) [reply endstate_correction/protocol.py:285]: this is not very verbose and is a potential point of failure or ambiguity.
There are two "sample pools" generated with two different Hamiltonians. Both of them are in the list. So `-1` is equivalent to `1`. but I will try to come up with a non-ambiguous way to solve this.

Comment (xiki-tempula) [reply endstate_correction/protocol.py:285]: I see. Maybe make this into a dictionary instead of a list `protocol.trajectories=dict(forward=AAA, backward=BBB)`?

File: endstate_correction/neq.py (hunk 2 above)
Comment (wiederm) [reply endstate_correction/neq.py:58]: oh, that is bug :-) good catch!
it is exactly as you assumed, it should read `samples.unitcell_lenghts[random_frame]`

File: endstate_correction/protocol.py (hunk 4 above)
Comment (wiederm) [reply endstate_correction/protocol.py:285]: I will add separate fields in the dataclass for reference/target samples. That might be the most verbose approach.
Thanks for raising this.

File: endstate_correction/tests/test_sampling.py (hunk 6 above)
Comment (wiederm) [reply endstate_correction/tests/test_sampling.py]: I have changed the target branch, I hope this is now easier to separate.
Sorry for that!

**Comment by xiki-tempula** [comment]
I think the test is failing as there is some problem with finding the cuda. I suggest pin it to cudatoolkit==11.4.2? As openMM can have trouble with latest version of CUDA 11.8

File: endstate_correction/neq.py (hunk 2 above)
Comment (wiederm) [reply endstate_correction/neq.py]: has been adressed here:
https://github.com/wiederm/endstate_correction/commit/bc21f74810bc0a8fd47b43979d39a7241144e164

File: endstate_correction/protocol.py (hunk 5 above)
Comment (wiederm) [reply endstate_correction/protocol.py]: has been resolved in https://github.com/wiederm/endstate_correction/commit/ff080e2a50b5b2982b805c8d246ca9bf753c4edb

File: endstate_correction/neq.py (hunk 1 above)
Comment (wiederm) [reply endstate_correction/neq.py]: added doc string
the documentation is still lacking, I will reserve some time in the next days to update it

**Comment by wiederm** [comment]
thanks! Yes, `direction` has been removed as a keyword. The direction is implicitly defined through the new keywords: `reference_samples` and `target_samples` (and reference-potential at lamb=0, target-potential at lamb=1). 
Similarily, if switching is performed uni- or bidirectional is based on if only reference or both reference and target samples are provided.
